### PR TITLE
More reliably clean up more SafeHandle instances

### PR DESCRIPTION
**Pull request by stephentoub**
This calls Dispose on SafeHandles (or things wrapping SafeHandles) that were otherwise being left for finalization:
1. Some of these are fixing finalization happening even on success paths (typically where the implementation isn't disposing of something that directly or indirectly wraps a SafeHandle).
2. Some of these are fixing finalization happening for invalid SafeHandles.  Their IsInvalid is true, but they still get finalized.
3. Some of these are fixing tests to finalize less.  My goal with fixing the tests was to eliminate the noise in order to find instances of the other two cases.

These were found primarily via two means:
- Debug-instrumentation in SafeHandle to log when one is finalized.  This instrumentation is built into debug/checked builds of SafeHandle and requires setting the DEBUG_SAFEHANDLE_FINALIZATION environment variable to "1".
- Auditing use of SafeHandle.IsInvalid

There's a lot more that can be cleaned up using the SafeHandle instrumentation, but I'm pausing here for now.  The System.IO.Pipes, System.IO.FileSystem, and System.Security.Cryptography tests are clean on Windows (with the exception of a few tests that are purposefully leaving objects for finalization to test finalization code paths).

### CODE REVIEW

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EcDsa.ImportExport.cs [hunk 1]
```diff
@@ -84,8 +84,9 @@ internal static SafeEcKeyHandle EcKeyCreateByExplicitCurve(ECCurve curve)
 
             if (key == null || key.IsInvalid)
             {
+                Exception e = Interop.Crypto.CreateOpenSslCryptographicException();
```
Comment (AaronRobinsonMSFT) [inline src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EcDsa.ImportExport.cs:87]: This pattern is reoccurring, is disposing of the handling impacting the computed exception?

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs [hunk 2]
```diff
@@ -141,6 +141,7 @@ internal override void TryToReuse(PipeValueTaskSource source)
 
                 if (handle.IsInvalid)
                 {
+                    handle.Dispose();
                     throw Win32Marshal.GetExceptionForLastWin32Error();
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs:145]: Do we need to capture the exception prior to disposing, like in the many other earlier examples?

File: src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/HelpersWindows.cs [hunk 3]
```diff
@@ -53,11 +53,16 @@ public static SafeHandle GetMsgParamAsMemory(this SafeCryptMsgHandle hCryptMsg,
...
             if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, paramType, index, IntPtr.Zero, ref cbData))
+            {
                 throw Marshal.GetLastWin32Error().ToCryptographicException();
+            }
 
             SafeHandle pvData = SafeHeapAllocHandle.Alloc(cbData);
             if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, paramType, index, pvData.DangerousGetHandle(), ref cbData))
+            {
+                pvData.Dispose();
                 throw Marshal.GetLastWin32Error().ToCryptographicException();
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/HelpersWindows.cs:64]: Capture before throw pattern?

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EcDsa.ImportExport.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EcDsa.ImportExport.cs:87]: > This pattern is reoccurring, is disposing of the handling impacting the computed exception?

In some cases I believe it could.  OpenSSL, for example, works on a stack of error codes, and any operation (including resource release) could in theory push an error on to the stack; then when we retrieve the last one to throw the exception, we'd get the wrong result.

In general, executing arbitrary code between when an error is ambiently stored and subsequently retrieved seems like a recipe for sometimes getting it wrong.

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs:145]: Yup, thanks.

File: src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/HelpersWindows.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/HelpersWindows.cs:64]: Yup

File: src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs [hunk 4]
```diff
@@ -241,16 +241,6 @@ public async Task Read_BaseStreamSlowly()
...
 
-        [Theory]
-        [InlineData(CompressionMode.Compress)]
-        [InlineData(CompressionMode.Decompress)]
-        public void CanDisposeBaseStream(CompressionMode mode)
-        {
-            var ms = new MemoryStream();
-            using var compressor = CreateStream(ms, mode);
-            ms.Dispose(); // This would throw if this was invalid
-        }
```
Comment (stephentoub) [inline src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs:252]: This test was bogus.  It wasn't disposing of the compression stream, which meant it was actually testing the impact of disposing the base stream.  Changing it to dispose of the compression stream highlights that it does in fact fail with an ObjectDisposedException when trying to flush the compression stream as part of disposal (for BrotliStream it always fails because of the footer, and for the zlib-based streams it fails if there's any data to be flushed).

File: src/libraries/System.IO.FileSystem/tests/File/OpenHandle.cs [hunk 5]
```diff
@@ -21,7 +21,18 @@ protected override FileStream CreateFileStream(string path, FileMode mode, FileA
             => new FileStream(File.OpenHandle(path, mode, access), access);
 
         protected override FileStream CreateFileStream(string path, FileMode mode, FileAccess access, FileShare share, int bufferSize, FileOptions options)
-            => new FileStream(File.OpenHandle(path, mode, access, share, options), access, bufferSize, (options & FileOptions.Asynchronous) != 0);
+        {
+            SafeFileHandle handle = File.OpenHandle(path, mode, access, share, options);
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.IO.FileSystem/tests/File/OpenHandle.cs:25]: I debated that.  This is the only one that passes additional options into the FileStream that can end up failing tests.  But, I can make them consistent for consistency sake.

**Review by jkotas — APPROVED** [review]

File: src/libraries/Common/src/System/Net/ContextAwareResult.Windows.cs [hunk 6]
```diff
@@ -10,7 +10,6 @@ internal sealed partial class ContextAwareResult
     {
         private WindowsIdentity? _windowsIdentity;
 
-        // Security: We need an assert for a call into WindowsIdentity.GetCurrent.
```
Comment (bartonjs) [inline src/libraries/Common/src/System/Net/ContextAwareResult.Windows.cs:13]: This doesn't seem related

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/ContextAwareResult.Windows.cs:13]: It's related in that I was auditing use of GetCurrent, this showed up in the search, and the comment is erroneous.

File: src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs [hunk 7]
```diff
@@ -139,7 +139,7 @@ public virtual async Task Dispose_WithUnfinishedReadAsync()
 
             using (var readStream = await ManualSyncMemoryStream.GetStreamFromFileAsync(compressedPath, false))
             {
-                var decompressor = CreateStream(readStream, CompressionMode.Decompress, true);
+                using var decompressor = CreateStream(readStream, CompressionMode.Decompress, true);
```
Comment (bartonjs) [inline src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs:142]: Personally, I'm not on fan of using braceless usings anywhere (including tests).

Just thought I'd throw it out there.

Comment (bartonjs) [reply src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs:142]: Not true. I got convinced that array-like pattern matching was better than hand writing the equivalent.  And `?.` is pretty good, too :).

**Review by bartonjs — APPROVED** [review]
LGTM, other than braceless usings in product code (which I believed we had banned in style outside of tests, but I guess not?)

**Comment by stephentoub** [comment]
> other than braceless usings in product code (which I believed we had banned in style outside of tests

I don't know why we'd ban them.

**Comment by danmoseley** [comment]
Yay, glad you added this switch. Do you plan to make an "up for grabs" issue to continue, in case folks are interested? eg., with a checkbox for each library that has hits (if you have that data gathered)?

Can you imagine us getting clean - I'm guessing there are test cases for finalizers, though.

How much slower is it to run tests when the switch is flipped? I assume it's expensive to gather stacks.

**Comment by stephentoub** [comment]
> How much slower is it to run tests when the switch is flipped? I assume it's expensive to gather stacks.

A lot... starting with the fact that it's only built into a debug/checked runtime, and then layering on a stack walk for every safe handle created.

> Can you imagine us getting clean - I'm guessing there are test cases for finalizers, though.

Without workarounds that'd be challenging, e.g. because of tests for finalizers as you say, tests that purposefully allow resources to not be disposed to avoid race conditions, etc.

> if you have that data gathered

I don't.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by danmoseley** [comment]
Makes sense -- probably we can use this flag on rare occasions, eg., having written a new feature with non trivial use of safehandles.

**Comment by bartonjs** [comment]
If you have to spin again, Steve, I'm indifferent about keeping or removing the version in https://github.com/dotnet/runtime/blob/388dd6c54c75ccad519ca094e3ddaa53a5c90d63/src/libraries/Common/src/Microsoft/Win32/SafeHandles/SafeX509Handles.Unix.cs#L12-L26 that I left in from getting Linux PKCS7 tidy.

**Comment by stephentoub** [comment]
> I'm indifferent about keeping or removing the version

Me, too. I will need to resolve a conflict, so I can remove it if you like.

**Comment by bartonjs** [comment]
Eh, if you're indifferent and I'm indifferent, let's leave it, so I can have a "smaller than the whole world" experience :)

**Comment by AndyAyersMS** [comment]
Possible regression: https://github.com/dotnet/perf-autofiling-issues/issues/6973

**Comment by bartonjs** [comment]
Since CryptoConfig.CreateFromName is just a lookup in ConcurrentDictionary, reflection to find a suitable ctor, and then reflection invocation of said ctor, I don't think this change was involved.

**Comment by stephentoub** [comment]
@bartonjs, I did touch this:
https://github.com/dotnet/runtime/pull/71991/files#diff-b5e1d8d2d47fbe472dd7de0d3d5450bc58de31363bee3ae108cf295bf879b4e4R44
I guess that could have caused this?

**Comment by bartonjs** [comment]
Yeah, I had that as a suspect, but then looked at the perf test:

https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/libraries/System.Security.Cryptography/Perf.CryptoConfig.cs

It goes straight to https://source.dot.net/#System.Security.Cryptography/System/Security/Cryptography/CryptoConfig.cs,2ea13b091ee4b642, not through the generic wrapper.

**Comment by stephentoub** [comment]
In that case, yeah, I'm not on the hook :)
